src/foam/nanos/crunch/JavaCrunchService.java

```diff
@@ -0,0 +1,140 @@
+package foam.nanos.crunch;
+
+import foam.core.X;
+import foam.dao.ArraySink;
+import foam.dao.DAO;
+import foam.mlang.predicate.Predicate;
+import foam.nanos.auth.Subject;
+import foam.nanos.auth.User;
+import foam.nanos.logger.Logger;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import static foam.mlang.MLang.*;
+
+public class JavaCrunchService implements CrunchService {
+  public List getGrantPath(X x, String rootId) {
+    Logger logger = (Logger) x.get("logger");
+
+    DAO prerequisiteDAO = (DAO) x.get("prerequisiteCapabilityJunctionDAO");
+    DAO capabilityDAO = (DAO) x.get("capabilityDAO");
+
+    // Lookup for indices of previously observed capabilities
+    Map<String,Integer> alreadyListed = new HashMap<String,Integer>();
+    // Return list (list is reversed at the end to put prerequisites first)
+    List grantPath = new ArrayList<>();
+
+    Queue<String> nextSources = new ArrayDeque<String>();
+    nextSources.add(rootId);
+
+    // Doing this instead of "this" could prevent unexpected behaviour
+    // in the incident CrunchService's getJunction method is decorated
+    CrunchService crunchService = (CrunchService) x.get("crunchService");
```

Comment:
> If the getJunction method was decorated on the crunchService - crunchService will use the decorated method.

I don't think this is true. If this CrunchService was a delegate of another CrunchService (extends ProxyCrunchService) that overrides getJunction, calling `this.getJunction` in this CrunchService will not call the decorated method.

Reply:
I was thinking that from the API user's perspective getGrantPath should always be consistent with getJunction, so if the getJunction method was overridden by a decorator then the base CrunchService should use that.

Although I'm willing to change it out for `this` if you think that's better.

Reply:
It's up to you. Your comment says that you don't want to use a decorated getJunction - "Doing this instead of "this" could prevent unexpected behaviour in the incident CrunchService's getJunction method is decorated" From what I understand, you're trying to avoid using decorated getJunction methods.

Reply:
oh, no exactly the opposite. If somebody overrides getJunction I want that change to be "noticed" by the getGrantPath method